cmd/dotmesh-server/s3_handlers.go

```diff
@@ -205,6 +208,71 @@ func (s *S3Handler) mountFilesystemSnapshot(filesystemId string, snapshotId stri
 	return e
 }
 
+func (s *S3Handler) headFile(resp http.ResponseWriter, req *http.Request, filesystemId, snapshotId, filename string) {
+	user := auth.GetUserFromCtx(req.Context())
+	fsm, err := s.state.InitFilesystemMachine(filesystemId)
+	if err != nil {
+		http.Error(resp, "failed to initialize filesystem", http.StatusInternalServerError)
+		return
+	}
+
+	state := fsm.GetCurrentState()
+	if state != "active" {
+		http.Error(resp, fmt.Sprintf("please try again later, state was %s", state), http.StatusServiceUnavailable)
+		return
+	}
+
+	// we must first mount the given snapshot before we try to read a file within it
+	// if snapshotId is not given then the latest snapshot id will be used
+	e := s.mountFilesystemSnapshot(filesystemId, snapshotId)
+	// the snapshot has been mounted - pass the SnapshotMountPath via the
+	// OutputFile to the fileOutputIO channel to get handled
+	if e.Name == "mounted" {
+		resp.Header().Set("Access-Control-Allow-Origin", "*")
+		resp.Header().Set("Content-Disposition", "attachment; filename=\""+filename+"\"")
+
+		respCh := make(chan *Event)
+		fsm.StatFile(&types.StatFile{
+			Filename:          filename,
+			User:              user.Name,
+			Response:          respCh,
+			SnapshotMountPath: (*e.Args)["mount-path"].(string),
+		})
+
+		result := <-respCh
+
+
+		switch result.Name {
+		case types.EventNameReadFailed:
+			err := result.Error()
+			if err != nil {
+				http.Error(resp, err.Error(), 500)
+				return
+			}
+			http.Error(resp, "read failed, could not retrieve actual error", 500)
+		case types.EventNameFileNotFound:
+			err := result.Error()
+			if err != nil {
+				http.Error(resp, err.Error(), 404)
+				return
+			}
+			http.Error(resp, "file not found, could not retrieve actual error", 404)
+		default:
+			if (*result.Args)["mode"].(os.FileMode).IsRegular() {
```

Comment:
would be good to check if it's set before converting to a type and calling a function or else it might panic :)